packages/dds/tree/src/feature-libraries/forest-summary/incrementalSummaryBuilder.ts

```diff
@@ -0,0 +1,430 @@
+			// parent's summary tree under its reference ID.
+			// Before encoding the chunk, set the parent summary builder to this chunk's summary builder so that
+			// any incremental chunks in the subtree of this chunk will use that as their parent summary builder.
+			const chunkSummaryBuilder = new SummaryTreeBuilder();
+			this.trackedSummaryProperties.parentSummaryBuilder = chunkSummaryBuilder;
+			chunkSummaryBuilder.addBlob(chunkContentsBlobKey, JSON.stringify(chunkEncoder(chunk)));
+
+			// Add this chunk's summary tree to the parent's summary tree. The summary tree contains its encoded
+			// contents and the summary trees of any incremental chunks under it.
+			parentSummaryBuilder.addWithStats(
+				`${chunkReferenceId}`,
+				chunkSummaryBuilder.getSummaryTree(),
+			);
+
+			// Restore the parent summary builder and chunk summary path.
+			this.trackedSummaryProperties.parentSummaryBuilder = parentSummaryBuilder;
+			this.trackedSummaryProperties.chunkSummaryPath.pop();
+		}
+
+		setInNestedMap(
+			this.chunkTrackingPropertiesMap,
+			this.trackedSummaryProperties.summarySequenceNumber,
+			chunk,
+			chunkProperties,
+		);
+		return [chunkReferenceId];
+	}
+
+	/**
+	 * Must be called after summary generation is complete to finish tracking the summary.
+	 * It clears any tracking state and deletes the tracking properties for summaries that are older than the
+	 * latest successful summary.
+	 * @param incrementalSummaryContext - The context for the incremental summary that contains the sequence numbers
+	 * for the current and latest summaries.
+	 */
+	public completedSummary(
+		incrementalSummaryContext: IExperimentalIncrementalSummaryContext | undefined,
+	): void {
+		if (!this.enableIncrementalSummary || incrementalSummaryContext === undefined) {
+			return;
+		}
+
+		validateTrackingSummary(this.forestSummaryState, this.trackedSummaryProperties);
+
+		// Delete tracking for summaries that are older than the latest successful summary because they will
+		// never be referenced again for generating summary handles.
+		this.chunkTrackingPropertiesMap.forEach((_, sequenceNumber) => {
+			if (sequenceNumber < this.latestSummarySequenceNumber) {
+				this.chunkTrackingPropertiesMap.delete(sequenceNumber);
+			}
+		});
+
+		this.forestSummaryState = ForestSummaryTrackingState.ReadyToTrack;
+		this.trackedSummaryProperties = undefined;
+	}
+
+	/**
+	 * Called to get the encoded contents of an incremental chunk with the given reference ID.
+	 * This is typically used when loading the forest to retrieve the contents of incremental chunks.
+	 * @param referenceId - The reference ID of the chunk to retrieve.
+	 * @returns The encoded contents of the chunk.
+	 */
+	public getEncodedIncrementalChunk(referenceId: ChunkReferenceId): EncodedFieldBatch {
+		const chunkEncodedContents = this.encodedChunkContentsMap.get(`${referenceId}`);
+		if (!chunkEncodedContents) {
+			throw new Error(`Chunk with reference ID ${referenceId} not found`);
```

Comment:
When possible, we should use errors that convey more information, like if this error is a UssageError of a bug in fluid (an assertion). I think this should be an assert.